### ISSUE TITLE
Transferring data from CPU to TPU takes too long if data structure is not same in all iterations

### ISSUE THREAD

**Issue by mfatih7** [pytorch/xla#5955]
Hi

I try to adapt my new learning setup to v2 and v3 TPUs on Google Cloud.
In each iteration, my training loop does not get the same data structure from the dataloader as shown below.

```
config.device = xm.xla_device()
...
...
for i, data in enumerate(dataloader):
	ii = i % (16 + 1 +16)
	if ii < 16:
		a_device = data['a'].to(config.device)
		...
		...	
	elif ii < (16 + 1):
		c_device = data['c'].to(config.device)
		d_device = data['d'].to(config.device)
		e_device = data['e'].to(config.device)
		f_device = data['f'].to(config.device)
		...
		...
	else ii < (16 + 1 +16):
		a_device = data['a'].to(config.device)
		b_device = data['b'].to(config.device)
		...
		...
...
...
```
While I run the learning loop line
`c_device = data['c'].to(config.device)`
takes too long to finish, more than a minute.
Should I preallocate a_device, b_device, c_device on TPU side?

I appreciate any help.

best regards

**Comment by mfatih7** [comment]
Ok. I understand that recompilation occurs since the structure of the input changes.
I will give a more representative code for my setup and ask if it possible to generate a model preventing recompilations.

**Comment by yeounoh** [comment]
For recompilation, would you be able to pad so that the shape is not dynamic? And to your original concern, the data loading during the training loop can actually block the execution, so we recommend using the `MpDeviceLoader` https://github.com/pytorch/xla/blob/7b92a9494914df25bf8e2db0d6a28def6297c167/test/test_train_mp_imagenet.py#L325 to overlap the data transfer with the execution.

**Comment by mfatih7** [comment]
Hi
Thank you for the answer.
Although using proper data loaders will positively impact the speed of the learning loop
The main reason for the slowness of my learning loop is the recompilations.

Here is the structure of my code. It works without any problem on GPUs.
As I mentioned I am trying to run the training loop on TPUs for faster training.

```
config.device = xm.xla_device()
N = 2048
batch_size = 64
batch_count_in_N = 16 ### 2048 /16
...
...
buffer_0 = torch.zeros( N, 1, dtype=torch.float32, requires_grad=False, device=config.device )
buffer_1 = torch.zeros( N, 1, dtype=torch.float32, requires_grad=True, device=config.device )
for i, data in enumerate(dataloader):
	ii = i % (16 + 1 +16)

## First Part
	if ii < 16:
		x_device = data['x'].to(config.device)
		
		outputs = model_A(x_device)
		
		## buffering outputs 
		buffer_0[ batch_size * ii : batch_size * ii + batch_size, : ] = outputs.detach()	

## Second Part
	elif ii < (16 + 1):
		label_0_device = data['label_0'].to(config.device)
		
		## Transferring buffer from a tensor requires_grad False to a tensor requires_grad True to enable backpropogation 
		buffer_1.data = buffer_0.data  
		
		loss_from_buffer = calculate_loss_wrt_c(buffer_1, label_0_device)
		
		## backpropogating to achieve buffer_1.grad
		loss_from_buffer.backward()

## Third Part
	else ii < (16 + 1 +16):
		x_device = data['x'].to(config.device)
		label_1_device = data['label_1'].to(config.device)
		
		outputs = model_A(x_device)
		
		feedback_from_buffer = buffer_1.grad[ batch_size*(ii-batch_count_in_N-1) : batch_size*(ii-batch_count_in_N-1)+batch_size ]
		labels_0_modified = outputs - config.learning_rate * feedback_from_buffer
		
		loss_0 = loss_function(labels_0_modified, outputs)
		loss_1 = loss_function(label_1_device, outputs)		
		loss_tot = loss_0 + loss_1
		
		## gradient accumulation
		loss_tot.backward()
		
		# Optimizing 
		if(ii==16+16):
			optimizer.step()
			optimizer.zero_grad()
		...
		...
...
...
```
The training loop has an input x and 2 labels as label_0, and label_1
The size of x is (batch_size, 2, 2048, 6) (B, C, H, W)
For each input, label_1 exists.
For each 2048 consecutive input, label_0 exists.
Therefore, to use both labels in loss calculations I need to train my model with a batch size of 2048.
However, when the input size is (2048, 2, 2048, 6) the activation maps in the network become too big to fit GPU and TPU memories.
Therefore on GPUs, I am using the above code with a batch size of 64 and gradient accumulation to fit the memories of the devices.
To optimize the network for 2048 input I implement the calculations in 3 parts.

In the first part, input(x) for batch sizes of 64 is pumped to the network 16 times and the output of the network is collected in buffer_0 which is a tensor with requires_grad False.
No backpropogation is applied in the first part since I need only the outputs for 2048 input to be used in calculations in second part.

In the second part, label_0 for 2048 input is taken from the data loader.
data in buffer_0 is transferred to buffer_1 which is a tensor with requires_grad True to enable back propagation.
Loss for buffer_1 is calculated and backpropogation is applied to get the grad values of buffer_1 to be used in the third part.

In the third part again, batches of 64 are pumped to the network as in part 1 but this time with label_1 values.
The output of the network is calculated for each batch of size 64.
Using the grad values of buffer_1 labels_0_modified is calculated.
labels_0_modified is another label for the training calculated using the label_0 in the second part.
loss_0 and loss_1 are calculated and summed.
The gradients are accumulated for loss_tot.
After loss_tot for all batches are backpropogated optimization is applied.

Is this training loop implementable on TPUs without recompilation?
I can move the accumulators and calculations into model_A.
If any part of this message is not clear I can explain.
